File: sort/quicksort.cpp
#include <iostream>
#include <vector>
using namespace std;

void quicksort(vector<int>& v, int begin, int end){
    if(end - begin<=1) return;
    int pivot = v[end-1], less_end = begin;

    for(int i=begin; i<end-1; i++)
        if(v[i]<pivot) swap(v[i], v[less_end++]);

    swap(v[end-1], v[less_end]);
    quicksort(v, begin, less_end);
    quicksort(v, less_end + 1, end);
}

int main()
{
	vector<int> vec = {7, 5, 16, 8};
	vec.push_back(26);
	quicksort(vec, 0, vec.size());
	for (int n : vec){
		cout << n << '\n';
	}
}

File: demo/sort/quicksort.py
def quicksort(a):
     if len(a) <= 1:
         return a  # 如果a為一位數則直接傳回a
     l = [x for x in a[1:] if x <= a[0]]  # 輸出排序後的比a[0]小的數列
     r = [x for x in a[1:] if x > a[0]]  # 輸出排序後的比a[0]大的數列
     return quicksort(l) + [a[0]] + quicksort(r)

print(quicksort([9,4,1,3,2,5,8,6,7]))
